### PR TITLE
Finally, arbitrary coord evaluation for swarms.

### PR DESCRIPTION
**Pull request by jmansour**
* Using nanoflann for kdtrees. Tests indicate this library is at least
  twice as fast as cKDTree. It's a C++ library.
* Each GeneralSwarm now records an index (the kdtree) for fast NN searching.
* Trees are only built/rebuilt when/if needed.
* Tree for 4M particles takes ~1s to build.
* Depending on query nature, querying 4M particles at 4M locations takes between 1s and 3s.
* Currently lack of caching means (svar+svar+svar).evaluate( (0.3,0.2) ) will do 3 queries for location.
  2 of these queries are redundant. There's a few options for improving on this which we can explore later.
* Currently all svar evaluations from element assembly will still use the old SwarmMap (PIC) method. Next
  step will be to consider switching this out.
* Function user guide updated with usage info and tests.

### CODE REVIEW

**Comment by lmoresi** [comment]
@jmansour, that seems like a good strategy and one that fixes the one issue that we have with Swarms. How about the other uses of k-d tree ( 👏  ) for SPR and distance values. Is the plan to add functionality at the py level for that using nanoflann ?  That's really my only question - what python wrapping work is needed for this to be used for nearest-particle and level-set type operations that we can currently do with the scipy version ?

**Comment by jmansour** [comment]
Yep. We can use `nanoflann` to collect a bunch of neighbours to construct RBF/SPR. `nanoflann` provides functionality for collecting the `k` nearest neighbours, or collecting all neighbours within a specified radius.  I haven't tested this yet however.  Not 100% sure yet, but we might like to create `RBF`/`SPR` filter functions.. something like this perhaps:
```python
swarm = uw.swarm.Swarm(mesh)
velgrad = vel.gradient[1]
rbf_velgrad = uw.function.filters.RBF(velgrad, swarm)
spr_velgrad = uw.function.filters.SPR(velgrad, mesh)
```
where we'd index the mesh just as we've indexed the swarms using kdtrees ( 👏 ) .  If I'm understand SPR correctly.   These would probably make most sense implemented at the C level, so have to look into what's involved with RBF in particular I guess.

**Comment by rbeucher** [comment]
Great Work @jmansour.

**Comment by jmansour** [comment]
Usage example at cell 64:
https://github.com/underworldcode/underworld2/blob/338f8992511a87da7602be27df5b7f1af7ae9917/docs/user_guide/04_Functions.ipynb

**Comment by lmoresi** [comment]
Is this always nearest neighbour ?  If we can provide an inverse-distance weighted interpolation then we could genuinely embrace the power of the k-d 👏 tree. I know this is just the first go, but it would be helpful to know if such an interface is possible.

**Comment by jmansour** [comment]
The library will return the list of NN particles (one or more), along with their distance from the query location.  This is all the information you require right?

**Comment by lmoresi** [comment]
Yep … I was wondering about providing additional args to the function evaluation / function definition that would specify nearest or N-nearest for the “interpolation”. That would open up quite a bit of flexibility in integration scheme development and testing for reducing the overall particle-load in models.

L

Prof Louis Moresi

louis.moresi@unimelb.edu.au<mailto:louis.moresi@unimelb.edu.au>

(w) +61 3 8344 1217

(m) +61 4 0333 1413

(us) +1 505 349 4425

www.moresi.info<http://www.moresi.info/>

www.facebook.com/underworldcode<http://www.facebook.com/underworldcode>

@LouisMoresi<https://twitter.com/LouisMoresi>


On 30 May 2019, 11:57 AM +1000, jmansour <notifications@github.com>, wrote:

The library will return the list of NN particles (one or more), along with their distance from the query location. This is all the information you require right?

**Comment by jmansour** [comment]
Yeah there's quite a few ways we can move forward with this, but we do need to be a bit careful to consider performance implications.  I haven't done the timing on it, but to obtain `k` nearest neighbours for a swarm of `n` particles scales as O(k log(n) ) apparently.  And then, we would multiply the function evaluations by `k` as well.  

We would I guess balance this increase in work by being able to use much lower particle densities.  But as importantly (or more) is the Gauss quadrature rule we would require.  Currently we default to 5^dim Gauss points (Q2).  In 3d this is 125 evaluations per element.  Use smarts such as weight approx would mean we can get away with much less particles while still sufficiently sampling the material swarm.   So for Q2, I guess we'd like to use only 3^dim Gauss points. 

There's no single correct way of doing things, and this does up the complexity of UW modelling, because it's no longer a vanilla swarm that behaves identically.  This is not a bad thing, but we will need to be careful to expose this in a sufficiently paternalistic 🙂 manner, and that we provide good documentation.

**Comment by lmoresi** [comment]
Why k function evaluations. If the call is something like swarm.evaluate(point, k=5) then the evaluations are all done at once at the C level aren’t they ?

Prof Louis Moresi

louis.moresi@unimelb.edu.au<mailto:louis.moresi@unimelb.edu.au>

(w) +61 3 8344 1217

(m) +61 4 0333 1413

(us) +1 505 349 4425

www.moresi.info<http://www.moresi.info/>

www.facebook.com/underworldcode<http://www.facebook.com/underworldcode>

@LouisMoresi<https://twitter.com/LouisMoresi>


On 30 May 2019, 12:45 PM +1000, jmansour <notifications@github.com>, wrote:

Yeah there's quite a few ways we can move forward with this, but we do need to be a bit careful to consider performance implications. I haven't done the timing on it, but to obtain k nearest neighbours for a swarm of n particles scales as O(k log(n) ) apparently. And then, we would multiply the function evaluations by k as well.

We would I guess balance this increase in work by being able to use much lower particle densities. But as importantly (or more) is the Gauss quadrature rule we would require. Currently we default to 5^dim Gauss points (Q2). In 3d this is 125 evaluations per element. Use smarts such as weight approx would mean we can get away with much less particles while still sufficiently sampling the material swarm. So for Q2, I guess we'd like to use only 3^dim Gauss points.

There's no single correct way of doing things, and this does up the complexity of UW modelling, because it's no longer a vanilla swarm that behaves identically. This is not a bad thing, but we will need to be careful to expose this in a sufficiently paternalistic 🙂 manner, and that we provide good documentation.

**Comment by jmansour** [comment]
Well it depends.  So I'm thinking of the filter approach:
```python
svar  = swarm.add_variable("double",1)
weighted_svar = uw.function.filter.inverse_weighted(swarm,svar, k=5)

weighted_svar.evaluate( (0.5,0.5) ) 
```
Here the filter would collect the neighbours, then extract the value from the provided `svar` function (just a swarm_variable), then perform the weighting.

However, it might make sense to do this
```python
mat_ind = swarm.add_variable('int',1)
mat_ind.data[::2] = 0
mat_ind.data[1::2] = 1

visc_dict = { 0: 1e1, 1: 1e5 }
visc_fn = uw.function.branching.map( mat_ind, visc_dict )

weighted_visc = uw.function.filter(swarm, visc_fn, k=5)

weighted_visc.evaluate( (0.5,0.5) )
```

Algorithmically there is nothing different here, except that instead of simply extracting 5 values from the `svar` swarm variable, the function evaluation now involves executing the branch function at 5 points before doing the weighting. 

Could easily get out of hand, but one positive is that it will scale trivially. 

Another option is limit the functionality by setting the behaviour on the swarm variable itself:
```python
svar.enable_weighted_average( k=5)

# but what the implications of doing this
(svar + svar + svar).evaluate( (0.5,0.5) )
```
Perhaps the best solution is to improve how we report back timings results. If the users have a clear understands of the implications of how they construct their functions, then it becomes less of an issue.

**Comment by lmoresi** [comment]
So I would suggest we also ask @brmather to comment on this. We have similar functions and requirements in quagmire ... but we also have some built in interpolants for unstructured data that can be near-neighbour, bilinear or cubic splines. It would be useful to think of a way to generalise this that can work in both environments. 

One point that occurs to me is that it would be helpful to be able to issue calls like this:

```python
temperatureField.evaluate((x,y,z), interpolant=Q3)

``` 

for the case where we need to have high order interpolation for a quantity (e.g. slade) but I am not sure about all the various ways that such choices would interact within complicated collections of functions which is presumably why you choose the filter. 

However, from the point of view of whether this PR is useful, this discussion is a sideline and I support the NN approach for the first implementation.

**Comment by jmansour** [comment]
Just for fun, here's the RT example where we've simply used two lines of particles to differentiate the materials. NN takes care of the rest. Note that this isn't currently possible with the PR, but I might add a parameter for the swarm so we can start playing with the dynamic possibilities of NN.  Apologies for bad scaling on vector arrows!

https://drive.google.com/open?id=1qSHIXK9StVIolePwBdkSgxQSr3n2sHGx

Here's the rms. Of course, this approach will only work up to a point. Also in parallel, with shadow space, I believe this will work fine. 

rt_vrms

**Comment by jmansour** [comment]
For the interface you suggest @lmoresi, I feel it would make sense just to create the filter on the fly:

```python
uw.functions.filter.interpolate(temperatureField, 'Q3').evaluate((x,y,z))
```
or similar.

**Comment by jmansour** [comment]
Here's another simulation, this time 64x64 (earlier was 128x128) and saner velocity arrows:

https://drive.google.com/open?id=1EVnwOlMx3_Hzc8OKJwD3UKQ6t_NqqtCs

Also, here's effectively how the materials appear to the matrix assembly due to NN:

https://drive.google.com/open?id=1TIs03q32jOVP9box57cnhrxUr3tqh0zL

and the vrms:

rt_vrms_64

**Comment by brmather** [comment]
Cool stuff @jmansour - just when we thought it could get any faster than the kdtree :clap: you find an even faster kdtree! :clap: :clap: 

In regards to the interface, we should allow users to specify the type of interpolation in Quagmire as in UW, but perhaps there ought to be a default for different data structures if it's unspecified by the user? E.g. nearest neighbour for querying points on the mesh; RBF for point swarms?

At least for Quagmire, I think we can make both interfaces possible:

```python
temperatureField.evaluate((x,y,z), 'Q3')
```

This is logical within the `lazyEvaluation` structure and makes `evaluate()` all-powerful :fist_raised: but perhaps at the expense of transparency (the input arguments are just `*args, **kwargs`), whereas

```python
fn.filter.interpolate(temperatureField, 'Q3').evaluate((x,y,z))
```

exposes all the possible interpolation arguments. In any case, sticking to a one-liner for interpolation is probably best and leave it to the user to determine which way they want to skin this cat.

**Comment by lmoresi** [comment]
@jmansour …

The example you show is a particle-based level set method which is very useful.

It might be helpful to implement a fn.level_set as we did in quagmire. This can be a useful replacement for complicated branching functions as well as for the kinds of things you have shown here.

Do you recommend using this library in quag instead of scipy ?

L

**Comment by jmansour** [comment]
I don't think it'd be too difficult to wrap. Might make sense to wrap it as an independent entity, using the `cKDTree` call patterns so that it can then act as a drop in replacement. Most of what I've read suggests it's generally the fastest library, but more testing would be useful.

Also, note that as discussed last week, the approaches discussed here give us an immediate solution to PETSc/UW3 function amenability, as all we ever use is the global locations. 

With regards to efficiency considerations for using multiple NN particles, my earlier concern was that we may be required to do a significant amount of redundant function evaluations. However, on further thought, I've realised that actually it would be trivial to cache and reuse the function results, ensuring that we never evaluate functions on particles more than once. This would require us to store in memory the results across the entire swarm population, but in the scheme of things that shouldn't be a problem.

**Comment by lmoresi** [comment]
@jmansour …

The example you show is a particle-based level set method which is very useful.

It might be helpful to implement a fn.level_set as we did in quagmire. This can be a useful replacement for complicated branching functions as well as for the kinds of things you have shown here.

Do you recommend using this library in quag instead of scipy ?



L


On 30 May 2019, 6:35 PM +1000, Ben Mather <notifications@github.com>, wrote:

Cool stuff @jmansour<https://github.com/jmansour> - just when we thought it could get any faster than the kdtree 👏 you find an even faster kdtree! 👏 👏

In regards to the interface, we should allow users to specify the type of interpolation in Quagmire as in UW, but perhaps there ought to be a default for different data structures if it's unspecified by the user? E.g. nearest neighbour for querying points on the mesh; RBF for point swarms?

At least for Quagmire, I think we can make both interfaces possible:

temperatureField.evaluate((x,y,z), 'Q3')

This is logical within the lazyEvaluation structure and makes evaluate() all-powerful ✊ but perhaps at the expense of transparency (the input arguments are just *args, **kwargs), whereas

fn.filter.interpolate(temperatureField, 'Q3').evaluate((x,y,z))

exposes all the possible interpolation arguments. In any case, sticking to a one-liner for interpolation is probably best and leave it to the user to determine which way they want to skin this cat.

**Comment by jmansour** [comment]
🤔 I thought I replied to this earlier, but now i can't find my reply.  Am I imagining things?

**Comment by jmansour** [comment]
@lmoresi just to reply to your earlier message (again!), I think `nanoflann` might be worth using. It would be _fun_ to write an API which mimics the `cKDTree` API so it can act as a drop in replacement for the wider community. However I probably wouldn't bother until `cKDTree` becomes a bottleneck.

Also, on further though, for weighted averaging methods (where we use `k` neighbours), it would be easy to cache particle evaluations to avoid redundant calculations. This way each particle would be evaluated at most once, with results recorded to an array of size `particleLocalCount`x`dim`.

**Comment by lmoresi** [comment]
I thought this page was useful … particularly the part with the higher-order interpolants:

https://www.xmswiki.com/wiki/WMS:Inverse_Distance_Weighted_Interpolation


Prof Louis Moresi

louis.moresi@unimelb.edu.au<mailto:louis.moresi@unimelb.edu.au>

(w) +61 3 8344 1217

(m) +61 4 0333 1413

(us) +1 505 349 4425

www.moresi.info<http://www.moresi.info/>

www.facebook.com/underworldcode<http://www.facebook.com/underworldcode>

@LouisMoresi<https://twitter.com/LouisMoresi>


On 7 Jun 2019, 4:22 PM +1000, jmansour <notifications@github.com>, wrote:

@lmoresi<https://github.com/lmoresi> just to reply to your earlier message (again!), I think nanoflann might be worth using. It would be fun to write an API which mimics the cKDTree API so it can act as a drop in replacement for the wider community. However I probably wouldn't bother until cKDTree becomes a bottleneck.

Also, on further though, for weighted averaging methods (where we use k neighbours), it would be easy to cache particle evaluations to avoid redundant calculations. This way each particle would be evaluated at most once, with results recorded to an array of size particleLocalCountxdim.

File: libUnderworld/PICellerator/MaterialPoints/src/GeneralSwarm.cpp [hunk 1]
```diff
@@ -434,3 +442,110 @@ void GeneralSwarm_ClearSwarmMaps( void* swarm ) {
     }
 }
 
+struct GeneralSwarm_nanoflann_interface
+{
+    GeneralSwarm* swarm;
```
Comment (julesghub) [inline libUnderworld/PICellerator/MaterialPoints/src/GeneralSwarm.cpp]: @jmansour I don't see the need for indents. If required consider use { ... } to wrap it.

Comment (jmansour) [reply libUnderworld/PICellerator/MaterialPoints/src/GeneralSwarm.cpp]: @julesghub, yep ew.  

invisibles on!  whitespace fixed!

**Comment by jmansour** [comment]
@julesghub there is an example, but I wasn't sure where it should live. Suggestions?

File: libUnderworld/PICellerator/MaterialPoints/src/GeneralSwarm.cpp (hunk 1 above)
Comment (julesghub) [reply libUnderworld/PICellerator/MaterialPoints/src/GeneralSwarm.cpp]: Wonderful!
Perhaps a gist for the working example, they can be made private i think.

File: libUnderworld/StgDomain/Swarm/src/SwarmVariable.h [hunk 2]
```diff
@@ -38,7 +38,8 @@
       double                                    magnitudeMin;  \
       double                                    magnitudeMax;  \
       Bool                                      useCacheMaxMin; \
-      Bool                                      addToSwarmParticleExtension;
+      Bool                                      useKDTree; \
+	  Bool                                      addToSwarmParticleExtension;
```
Comment (julesghub) [inline libUnderworld/StgDomain/Swarm/src/SwarmVariable.h]: look out!

Comment (julesghub) [inline libUnderworld/StgDomain/Swarm/src/SwarmVariable.h]: another white collar criminal

Comment (jmansour) [reply libUnderworld/StgDomain/Swarm/src/SwarmVariable.h]: I'm one step ahead 
4852c33

**Comment by julesghub** [comment]
@jmansour I tested in parallel, essentially repeating the operations of the user_guide/functions.ipynb. I call swarm.shadow_particles_fetch() but the shadow particles don't appear to be shared, so the NN result is not the same in serial. Is it parallel safe?
I can upload the test script if wanted.

**Comment by jmansour** [comment]
I'll upload the model here for now. 

[RTNN.tar.gz](https://github.com/underworldcode/underworld2/files/3299462/RTNN.tar.gz)

Note, this probably won't work even with shadow space, as procs aren't generally guaranteed to have any particles.  Ignore what I said earlier! 

I'll have a closer look at the `Functions.ipynb` examples in parallel.

**Comment by jmansour** [comment]
@julesghub  I've pushed up the changes.  Try and rerun your model to see. 

Note that shadow particles are not indexed by the kdtree.  I looked into it, and the required design isn't delightful, so I think it's better to just wait until we do the proper kdtree parallel schema.

**Comment by julesghub** [comment]
ok, tested the parallel error. Nice one Johnny!
For reference its `swarm.allow_parallel_nn=True`.

I'll re-run tests and if all good i'll merge the PR into dev.

**Comment by jmansour** [comment]
Thanks dude!

**Pull request merged**